working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+# Proposal for adding `FinalizationRegistry` and `WeakRef` to core libraries.
+
+Author: vegorov@google.com<br>Version: 0.1
+
+## Background
+
+This proposal includes two interconnected pieces of functionality:
+
+- ability to create _weak references_ (`WeakRef`);
+- ability to associate finalization actions with objects
+  (`FinalizationRegistry`).
+
+### Weak references
+
+_Weak reference_ is a object which holds to its _target_ in a manner which does
+not prevent runtime system from reclaiming the referenced object when it becomes
+_weakly reachable_ or, in other words, reachable only through weak references.
+When an object is dimmed _weakly reachable_ all `WeakRef` objects referencing it
+are cleared and the object is eligible for reclamation.
+
+### Finalization
+
+_Finalization_ is a process of invoking a piece of user defined code at some
+point after certain object has been deemed _unreachable_ by a garbage collector.
+_Unreachable_ objects can no longer be interacted with by any programmatic
+means and thus can be safely reclaimed in a manner otherwise unobservable
+to the rest of the program.
+
+Finalization provides a way to observe and react to the event of runtime system
+discovering that some object is unreachable and can be used in a variety of
+ways. Most commonly finalization is used to cleanup some sort of manually
+managed (native) resources associated with a GC managed objects.
+
+For example, finalization can be used to automatically free native resources
+accessed through `dart:ffi` when Dart objects owning these resources become
+unreachable.
+
+```dart
+/// Wrapper for a native resource (accessed through the [_resource] pointer).
+class Wrapper {
+  final ffi.Pointer<ffi.Void> _resource;
+
+  /// Without automatic finalization [Wrapper] user needs to explicitly
+  /// call [destroy] to free the underlying [_resource], when it is no
+  /// longer needed, leading to a possibility of leaks.
+  void destroy() {
+    malloc.free(_resource);
+  }
+}
+```
+
+### Prior art in Dart VM Embedding API
```

Comment:
Can you also briefly mention the prior art from the JavaScript implementation of WeakRef and FinalizationRegistry?